demux/demux.c

```diff
@@ -1990,7 +1990,7 @@ static void record_packet(struct demux_internal *in, struct demux_packet *dp)
 static void add_packet_locked(struct sh_stream *stream, demux_packet_t *dp)
 {
     struct demux_stream *ds = stream ? stream->ds : NULL;
-    if (!dp->len || demux_cancel_test(ds->in->d_thread)) {
+    if (!dp->len || !ds || demux_cancel_test(ds->in->d_thread)) {
```

Comment:
It would seem that:

* `ds` is always meant to be there (at least looking at the bunch of unchecked `ds` usage in this function that would otherwise (hopefully) crash & burn.
* `ds->in` has its contents being de-referenced in both the `demux_cancel_test` in the if, as well as right after the two asserts.

So it almost seems like something a la
```c
assert(ds && ds->in);
```

Would not be too much out of place just before the if as we probably would want to be notified if either `ds` or `ds->in` being nullptr would ever to happen?

Reply:
Not sure about assert, we have return value that can be used, so we just return on error.

Reply:
For me the difference in general is:

* things that should never happen and are definite errors (esp. in internal API usage) are quite open for asserts. In this case if this was happening, we should already be crashing, `ds` itself is supposed to be talloc'd in `demux_add_sh_stream_locked(`, as far as I can see.
** the call sites of course should be checked in these cases to confirm whether the case should be happening.
* things that may happen due to normal logic but can't handled (also often just error cases in external interfaces) should return an error.